netdb.c

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2017, [Ribose Inc](https://www.ribose.com).
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <arpa/inet.h>
+
+#include "common.h"
+#include "rtr-netdb.h"
+
+struct hostent *RETRACE_IMPLEMENTATION(gethostbyname)(const char *name)
+{
+	rtr_gethostbyname_t real_gethostbyname;
+	struct hostent *hent;
+
+	real_gethostbyname = RETRACE_GET_REAL(gethostbyname);
+
+	hent = real_gethostbyname(name);
+	if (hent) {
+		struct in_addr **addr_list;
+		int i;
+
+		trace_printf(1, "gethostbyname(\"%s\") = [", name);
+
+		addr_list = (struct in_addr **) hent->h_addr_list;
+		for (i = 0; addr_list[i] != NULL; i++) {
+			char ip_addr[INET6_ADDRSTRLEN];
+
+			inet_ntop(hent->h_addrtype, addr_list[i], ip_addr, sizeof(ip_addr));
+			if (i > 0)
+				trace_printf(0, ",");
+			trace_printf(0, "%s", ip_addr);
+		}
+
+		trace_printf(0, "]\n");
+	} else
+		trace_printf(1, "gethostbyname(\"%s\") = NULL [error:%s]\n", name, hstrerror(h_errno));
+
+	return hent;
+}
+
+RETRACE_REPLACE(gethostbyname);
+
+struct hostent *RETRACE_IMPLEMENTATION(gethostbyaddr)(const void *addr, socklen_t len, int type)
+{
+	rtr_gethostbyaddr_t real_gethostbyaddr;
+	struct hostent *hent;
+
+	real_gethostbyaddr = RETRACE_GET_REAL(gethostbyaddr);
+
+	hent = real_gethostbyaddr(addr, len, type);
+	if (addr) {
+		char ip_addr[INET6_ADDRSTRLEN];
+
+		if (type == AF_INET) {
+			struct in_addr *in_addr = (struct in_addr *) addr;
+
+			inet_ntop(AF_INET, in_addr, ip_addr, sizeof(ip_addr));
+		} else if (type == AF_INET6) {
+			struct in6_addr *in6_addr = (struct in6_addr *) addr;
+
+			inet_ntop(AF_INET, in6_addr, ip_addr, sizeof(ip_addr));
```

Comment:
AF_INET6` ? Do you need `if` `else` here? Would plain `inet_ntop(AF_INET, addr, ip_addr, sizeof(ip_addr));` do?